www/js/lib/zstdec_wrapper.js

```diff
@@ -0,0 +1,265 @@
+     * @property {Integer} _inStreamChunkedPos The position once the currently loaded chunk will have been consumed
+     * @property {Integer} _outStreamPos The position in the decoded byte stream (offset from start of cluster)
+     * @property {Array} _outDataBuf The buffer that stores decoded bytes (it is set to the requested blob's length, and when full, the data are returned)
+     * @property {Integer} _outDataBufPos The number of bytes of the requested blob decoded so far
+     */
+    
+    /**
+     * @constructor
+     * @param {FileReader} reader The reader used to extract file slices (defined in zimfile.js)
+     */
+    function Decompressor(reader) {
+        this._reader = reader;
+    }
+
+    /**
+     * Set up the decompression stream, and initiate a read loop to decompress from the beginning of the cluster
+     * until we reach <offset> in the decompressed byte stream
+     * @param {Integer} offset Cluster offset (in deocmpressed stream) from which to start reading
+     * @param {Integer} length Number of decompressed bytes to read
+     * @returns {Promise<ArrayBuffer>} Promise for an ArrayBuffer with decoded data
+     */
+    Decompressor.prototype.readSlice = function(offset, length) {
+        busy = true;
+        this._inStreamPos = 0;
+        this._inStreamChunkedPos = 0;
+        this._outStreamPos = 0;
+        this._outDataBuf = new Int8Array(new ArrayBuffer(length));
+        this._outDataBufPos = 0;
+        var ret = zd._ZSTD_initDStream(zd._decHandle);
+        if (zd._ZSTD_isError(ret)) {
+            return Q.reject('Failed to initialize ZSTD decompression');
+        }
+
+        return this._readLoop(offset, length).then(function(data) {
+            // DEV: We are re-using all the allocated w/asm memory, so we do not need to free any of structures assigned wiht _malloc
+            // However, should you need to free assigned structures use, e.g., zd._free(zd._inBuffer.src);
+            // Additionally, freeing zd._decHandle is not needed, and actually increases memory consumption (crashing zstddeclib)
+            // Should you need to free the decoder stream handle, use command below, but be sure to create a new stream control object
+            // before attempting further decompression
+            // zd._ZSTD_freeDStream(zd._decHandle);
+            busy = false;
+            return data;
+        });
+    };
+
+    /**
+     * This function ensures that only one decompression runs at a time, launching readSlice() only when
+     * the decompressor is no longer busy
+     * @param {Integer} offset The cluster offset (in decompressed stream) at which the requested blob resides
+     * @param {Integer} length The number of decompressed bytes to read
+     * @returns {Promise} A Promise for the readSlice() function
+     */
+    Decompressor.prototype.readSliceSingleThread = function (offset, length) {
```

Comment:
This Single Thread feature was necessary with xz decoder to keep memory consumption reasonable (#389). Is it also necessary for zstd? (it might be considered as an enhancement, and tested later)

Reply:
It is necessary because we are not setting up new memory structures, but are explicitly re-using the same ones in the Emscripten HEAP as suggested in the C++ source code. This saves a lot of memory. I did actually try to eliminate this, but just as with XZ, the issue is that if we want to have two or more decompressions happening simultaneously, they each need their own memory structures and demand double/triple/quadruple the memory. We could potentially have dozens running. Of course none of them really are multi-threaded in JS, unless we launch a different WORKER for each decompression, but we don't have that code. Promises are only the illusion of multi-threading, after all...

Reply:
I understand. So we have the same issue as with xz and need to keep it.
Thanks a lot for the explanation